R/fair-aaa.R

```diff
@@ -0,0 +1,137 @@
+#' Create fairness metrics
+#'
+#' Fairness metrics quantify the disparity in value of a metric across a number
+#' of groups. Fairness metrics with a value of zero indicate that the
+#' underlying metric has parity across groups. yardstick defines
+#' several common fairness metrics using this function, such as
+#' [demographic_parity()], [equal_opportunity()], and [equalized_odds()].
+#'
+#' @param .fn A yardstick metric function or metric set.
+#' @param .name The name of the metric to place in the `.metric` column
+#' of the output.
+#' @param .post A function to post-process the generated metric set results `x`.
+#' In many cases, `~diff(range(x$.estimate))` or
+#' `~r <- range(x$.estimate); r[1]/r[2]`.
+#'
+#' @section Relevant Group Level:
+#' By default,
+#'
+#' Additional arguments can be passed to the function outputted by
+#' the function that this function outputs. That is:
+#'
+#' ```
+#' res_fairness <- fairness_metric(...)
+#' res_by <- res_fairness(by)
+#' res_by(..., additional_arguments_to_.post = TRUE)
+#' ```
+#'
+#' For finer control of how groups in `by` are treated, use the
+#' `.post` argument.
+#'
+#' @return A function with one argument, `by`, indicating the data-masked
+#' variable giving the sensitive feature. See the documentation on the
+#' return value of fairness metrics like [demographic_parity()],
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
+#'
+#' # `demographic_parity`, among other fairness metrics,
+#' # is generated with `fairness_metric()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#' demographic_parity_ <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity",
+#'     .post = diff_range
+#'   )
+#'
+#' m_set <- metric_set(demographic_parity_(Resample))
+#'
+#' m_set(hpc_cv, truth = obs, estimate = pred)
+#'
+#' # the `post` argument can be used to accommodate a wide
+#' # variety of parameterizations. to encode demographic
+#' # parity as a ratio inside of a difference, for example:
+#' ratio_range <- function(x, ...) {
+#'   range <- range(x$.estimate)
+#'   range[1] / range[2]
+#' }
+#'
+#' demographic_parity_ratio <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity_ratio",
+#'     .post = ratio_range
+#'   )
+#'
+#' @export
+fairness_metric <- function(.fn, .name, .post) {
```

Comment:
Notes from the group meeting:
* Max mentioned that it might be nice to prefix whatever this function name is with `create_` or some other eliciting verb to indicate that this is a function factory, and others agreed
* I suggested `disparity_metric` as a descriptor for this type of metric that doesn't have as strong of a social connotation—seems like "disparity" could describe differences across groups regardless of whether that group is regarded as a sensitive feature

Reply:
I like the idea of using `create_*` here, but it is a departure compared to the other function factories in yardstick (of which there are a lot, like `metric_tweak()`, `metric_set()`, and so forth). Do you think it's better to stay more similar to the naming conventions of yardstick, or to use something like `create_*`?

I have a mild preference for something like `create_groupwise_metric()` because I think there is more ML community vocabulary around what "groupwise" means. The word "disparity" makes me think about the specific metric [disparate impact](https://www.ibm.com/docs/en/cloud-paks/cp-data/4.5.x?topic=monitor-fairness-metric-disparate-impact). That being said, my opinion is not super strong here.

Reply:
Good points--if we want to look to other function factories in the package, maybe the parallel we might want to draw is with `new_metric()`? Something like `new_groupwise_metric()`?

Reply:
I like that a lot, `new_groupwise_metric()` 👍